Make `wgpu_hal::api::Empty` into a usable `Noop` backend for testing.

kpreid opened the pull request:
**Connections**
Prototype implementation of #5236.

**Description**
This allows users to create a nonfunctional `wgpu` instance by setting options to use the `Noop` backend (formerly `Empty), which allows creating resources and issuing commands (that will never run) without having any actual GPU or other device available. This is intended to be used in unit tests for code which manages `wgpu` resources. It does so by exposing a safe means to create `wgpu_hal::api::Empty`, and by tweaking the behavior of `Empty` slightly so that it allows more things to succeed.

Resolved questions: ~~This is a draft because I am unsure about many elements and want to get feedback before I put more work into it. Particularly:~~

* ~~It seems to me not ideal that this depends on `wgpu-core` — but `Empty` was already there. I was originally planning to write a `wgpu` backend instead, but this seemed like a simpler place to start. If we like the direction this goes, then perhaps it would be even better to have a `wgpu` backend so there is no need for a feature and no need to build `wgpu-core` (e.g. for an application which uses WebGPU only).~~
* ~~To get it to run, I had to add a `Backends` bit for `Empty`, which means that `Backends::all()` includes `Empty`. But applications which ask for `Backends::all()` will then (semantically if not in practice) be saying they will accept a nonfunctional implementation.~~ It is now necessary to specify a special option.
* ~~Are there any caveats to using `Empty` this way?~~
    * ~~It seems to be only used for some tests, so perhaps exposing it is incorrect/unsound in some way.~~
    * ~~Will making more of its operations succeed rather than fail break any of its existing usage?~~ No.
* ~~I named the public feature “noop” instead of “empty”, but probably there should be consistent naming — I just don't think `Empty` is a good public name.~~ Renamed all to `Noop`.
* ~~Ideally the noop device would accept all commands, but become “lost” immediately upon any action that would be able to view the output (i.e. mapping a buffer for read). I don’t yet know how to do this precisely.~~ The noop device now allows buffer manipulation but nothing else.

**Testing**
Added a test function demonstrating minimum functionality. No manual testing or specific hardware required to test! That's the whole point!

<!--
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're no bothering us!
-->

**Checklist**

- [X] Run `cargo fmt`.
- [ ] Run `taplo format`.
- [X] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
- [X] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

sagudev commented:
> * It seems to me not ideal that this depends on `wgpu-core` — but `Empty` was already there. I was originally planning to write a `wgpu` backend instead, but this seemed like a simpler place to start. If we like the direction this goes, then perhaps it would be even better to have a `wgpu` backend so there is no need for a feature and no need to build `wgpu-core` (e.g. for an application which uses WebGPU only).

Most of validation is in wgpu-core anyway, so I am not sure how much useful would empty wgpu backend be.

kpreid commented:
> Most of validation is in wgpu-core anyway, so I am not sure how much useful would empty wgpu backend be.

Thank you for that clarification; I wasn’t sure if that was true, and it is quite important for some testing purposes that validation *is* performed.

cwfitzgerald commented on wgpu-hal/src/empty.rs:
```diff
@@ -83,7 +87,30 @@ impl crate::Instance for Context {
+                driver_info: String::new(),
+                backend: wgt::Backend::Empty,
+            },
+            features: wgt::Features::default(),
```
This should be `all()`

kpreid replied:
Done.

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -146,6 +146,9 @@ bitflags::bitflags! {
     #[cfg_attr(feature = "serde", serde(transparent))]
     #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
     pub struct Backends: u32 {
+        /// TODO: can't live with this
```
Honestly I think this is fine. Will explain more in my top level comment.

kpreid replied:
I added `NoopBackendOptions` as you suggested.

cwfitzgerald commented on wgpu/Cargo.toml:
```diff
@@ -57,6 +61,12 @@ vulkan-portability = ["wgpu-core?/vulkan"]
 ## * ⚠️ WIP: Currently will also enable GLES dependencies on any other targets.
 webgl = ["dep:wgpu-hal", "wgpu-core/gles"]
 
+## Enables the noop backend for testing.
+##
+## This backend allows creating resources such as buffers and textures,
+## but performs no computation.
+noop = ["dep:wgpu-core"]
```
I think this is fine as a backend feature (especially after #7076). Though we probably should make `noop` a feature at the wgpu-hal level as well, cutting off the empty backend if it's not needed.

kpreid replied:
There are some comments in `wgpu_core::id::Id` suggesting it is used as a placeholder type — I can't actually find that usage, though. Perhaps the comments are stale.

cwfitzgerald replied:
yeah I think those have been removed as part of the arcanization. Needn't hold up this PR on paranoia anyway.

kpreid replied:
I think I'd like to keep this PR smaller by not *adding feature conditionals* to existing code (a fragile process) as part of it.

cwfitzgerald replied:
Yup, all good.

cwfitzgerald commented on wgpu/Cargo.toml:
```diff
@@ -28,6 +28,10 @@ ignored = ["cfg_aliases"]
 
 [lib]
 
+[[test]]
+name = "no-gpu"
+required-features = ["noop"]
+
```
I find `required-features` really annoying (need to remember to enable them). Lets make a `/tests/validation_tests` or something, which is a integration test part of the wgpu-test library, and can enable all the features it needs.

kpreid replied:
Done.

cwfitzgerald reviewed (changes requested):
This is so cool and I've been wanting to have this for a very long time!

I think we can handle the Backends bitflag like this:
- We include the `Empty` backend as in this PR.
- We add a `EmptyBackendOptions` which has a single boolean `enable`, which defaults false.

This way the empty backend is "enabled", but doesn't actually return any adapter until the enable boolean is set.

Though now thinking about it, if this is feature gated anyway, does it matter? Maybe it does because of feature unification?

cwfitzgerald commented:
> It seems to be only used for some tests, so perhaps exposing it is incorrect/unsound in some way.

We just never wired it up properly.

> Will making more of its operations succeed rather than fail break any of its existing usage?

I'm unaware of any, and if it's there, we can break it, this wasn't a feature we advertised.

> I named the public feature “noop” instead of “empty”, but probably there should be consistent naming — I just don't think Empty is a good public name.

Noop backend is a pretty good name, I think changing is is a-okay

>  Ideally the noop device would accept all commands, but become “lost” immediately upon any action that would be able to view the output (i.e. mapping a buffer for read). I don’t yet know how to do this precisely.

I wouldn't worry about it, I think just always returning zeros (or whatever data you wrote it) for mappings is okay.

kpreid committed:
Rename `Empty` backend to `Noop`.

This is in preparation for making it a more substantial test stub,
able to execute nontrivial code, and usable from the safe `wgpu` API.

kpreid committed:
Extend the capabilities of the `Noop` backend.

This will allow using it in tests of wgpu resource management code
that does not actually require a backend.

* `enumerate_adapters()` returns an adapter instead of failing.
* `open()` returns a device instead of failing.
* `create_buffer()` allocates actual memory.
* `map_buffer()` actually provides access to that memory.
* `clear_buffer()` actually clears the buffer.
* `copy_buffer_to_buffer()` actually copies data.
* Fences actually work (trivially, because all operations are
  synchronous).

Future work could include implementing texture copies,
timestamp queries, and the clearing part of render passes.

kpreid committed:
Allow creating `Noop` backend through the normal mechanisms.

kpreid committed:
Add a test for the noop backend.

kpreid commented:
> Though now thinking about it, if this is feature gated anyway, does it matter? Maybe it does because of feature unification?

Indeed, we need to not return a noop backend just because some other package, which is either uncareful about minimal features or has a non-test-only use case for it, has enabled the feature.

> Noop backend is a pretty good name, I think changing is is a-okay

Done.

> I wouldn't worry about it, I think just always returning zeros (or whatever data you wrote it) for mappings is okay.

I've now added support for copying and reading buffers, because that seemed likely to be valuable — it will allow full testing of any module/middleware whose job is “put things in a buffer in the proper layout” — and because having it helps prove that the buffer mapping code I need anyway is actually correct; the previous iteration of this PR was actually unsound.

Texture support would also be very useful, but significantly harder and so I didn't include it in this PR.

cwfitzgerald commented on tests/Cargo.toml:
```diff
@@ -22,11 +22,16 @@ name = "wgpu-compile-test"
 
 [dependencies]
-wgpu.workspace = true
+wgpu = { workspace = true, features = ["noop"] }
```
This is a non-obvious one, we should move the feature here to the workspace - I've unified all test uses to be straight workspace deps, and it's not a real harm for tests

kpreid replied:
That seems like it would be bad if the workspace ever gains any non-test packages that depend on `wgpu`. (Arguably, `wgpu-info` is one already, though it would be reasonable for it to include noop.)

cwfitzgerald replied:
Potentially - I made that change in part because we kept forgetting features between test/wgpu-info (which need to be in sync) and other areas.

I think it's fine for now though.

kpreid replied:
Done.

cwfitzgerald commented on tests/src/init.rs:
```diff
@@ -43,6 +43,10 @@ pub fn initialize_instance(backends: wgpu::Backends, force_fxc: bool) -> Instanc
                 shader_compiler: dx12_shader_compiler,
             },
             gl: wgpu::GlBackendOptions::from_env_or_default(),
+            // TODO: Use noop backend in tests instead of leaving it disabled by default?
```
Could this be an issue? Todos aren't terribly visible

kpreid replied:
https://github.com/gfx-rs/wgpu/issues/7119

cwfitzgerald commented on wgpu-core/Cargo.toml:
```diff
@@ -113,6 +113,9 @@ gles = ["wgpu-hal/gles"]
 dx12 = ["wgpu-hal/dx12"]
 
+## Enable the `noop` backend.
+noop = [] # currently not conditional in wgpu-hal
```
Issue pls

kpreid replied:
https://github.com/gfx-rs/wgpu/issues/7120

cwfitzgerald commented on wgpu-hal/src/noop/buffer.rs:
```diff
@@ -0,0 +1,82 @@
+unsafe impl Sync for Buffer {}
+
+impl Buffer {
+    pub(in crate::noop) fn new(desc: &crate::BufferDescriptor) -> Result<Self, crate::DeviceError> {
```
```suggestion
    pub(super) fn new(desc: &crate::BufferDescriptor) -> Result<Self, crate::DeviceError> {
```

Convention here is just pub(super)

kpreid replied:
IMO, using `super`, particularly across files, is fragile and best kept to a minimum, because it means that module-moving refactors can have surprising effects. I’ll change it anyway, but I am noting that I disagree with this policy.

cwfitzgerald replied:
Yeah that makes sense - I don't really have much of a preference either way, just following existing convention. Might be worth an issue about changing it. While hal backends have historically been a single mod + submodules, as we split things up more, we might see deeper splits.

kpreid replied:
[Done.](https://github.com/gfx-rs/wgpu/issues/7121)

cwfitzgerald commented on wgpu-hal/src/noop/mod.rs:
```diff
@@ -83,7 +109,31 @@ impl crate::Instance for Context {
+            },
+            features: wgt::Features::all(),
+            capabilities: crate::Capabilities {
+                limits: wgt::Limits::default(),
```
This should probably be spelled out as something high. Potentially just u32::MAX/1 depending on max/min limit type.

kpreid replied:
Done.

cwfitzgerald reviewed (approved):
Incredible work as always!

Some comments, but all minor.

cwfitzgerald commented:
This just needs conflicts fixed then we're good to go.

kpreid commented:
Whoops, I rebased but without having fetched today. Should be good to merge now.

kpreid commented:
@cwfitzgerald My pushing disabled auto-merge so you're going to have to hit the button again.

…also I see you're planning to squash this, so I had better pre-squash it so the commit messages aren't lost. Grumble grumble.

cwfitzgerald commented:
I can rebase, I just wasn't sure if they were standalone

kpreid commented:
All the commits I previously pushed were self-contained and (should) pass tests. Shall I restore them? Right now I have squashed to one commit, but I don't prefer that.

kpreid commented:
I have now re-pushed the version with incremental, informative commits.

cwfitzgerald commented:
Yeah just let us know when we can rebase and merge, and we will. We're going to add this to the pull request template shortly.
